### ISSUE TITLE
Metamodel declaration.name inconsistency between Java and JavaScript

### ISSUE THREAD

**Issue by lukedegruchy** [ceylon/ceylon.language#668]
Not sure if this bug belongs in ceylon.language.  

1) Create an Eclipse project that compiles to Java and JavaScript
2) Create the following file:

```ceylon
import ceylon.language.meta.model {
    Type,
    ClassOrInterface
}

shared void run() {
    class SomeSub() extends Some<SomeSub>() {}

    print(SomeSub().name);
}

shared class Some<ThisType>() {
    shared default String name {
        Type<ThisType> thisType = `ThisType`;
        
        assert(is ClassOrInterface<ThisType> thisType);
        return thisType.declaration.name;
    }
}
```

3) Compile and run it in both Java and JavaScript  (same results on Eclipse and command line):

cd ${ECLIPSE_PROJECT_ROOT}
$ ${PATH_TO_CEYLON_GITHUB}/ceylon-dist/dist/bin/ceylon compile
Note: Created module luke.metamodel/1.0.0
$ ${PATH_TO_CEYLON_GITHUB}/ceylon-dist/dist/bin/ceylon compile-js
Note: Created module luke.metamodel/1.0.0
$ ${PATH_TO_CEYLON_GITHUB}/ceylon-dist/dist/bin/ceylon run luke.metamodel/1.0.0
SomeSub
$ ${PATH_TO_CEYLON_GITHUB}/ceylon-dist/dist/bin/ceylon run-js luke.metamodel/1.0.0
SomeSub$qi7q8r

So basically, the output in Java is "SomeSub" and the output in JavaScript is "SomeSub$qi7q8r".

**Comment by gavinking** [comment]
Yours, @chochos :)
